### ISSUE TITLE
Precaching is really slow (and not because the resources are too large)

### ISSUE THREAD

**Issue by boerwastaken** [catalpainternational/canoe#98]
Preloading resources is really slow on Asteroid. It looks like it's sequential and it looks like it could be a lot better. It also looks like 

I compared with Eskola/Gather to back this up. Measured today on edutest.temp.build vs asteroid.tl in private mode on empty caches empty everything. See notes for some details.

## Results

|         | Eskola           | Asteroid  | Factor |
| ------------- |-------------:| -----:| ----:|
| Load time (s)      | 25 | 62 | 0.40 |
| \# requests      | 237      |  153 | 1.55
| total xferd (kB) | 7.5  | 2.6 | 2.88

That's a pretty spectacular difference. Gather's preload loads **2.88 × more data**, spread over **1.55 × more requests**, in **40% of the time** it takes Asteroid's preloader.

It's even more baffling when you realize there are a lot of things working in Canoe's favour: 

- Eskola only installs the serviceworker *after* the login, so that takes wall clock time, clearly visible at the start
- In the preload Asteroid loads **30** resources through the Django API, which is much slower (notably latency) than just getting a static file through nginx. But it's not as if Eskola is just happily whirling in only static resources, on the contrary: it hits Django **60** times, so twice as much, and those resources come from some pretty complex queries, five take more than a second TTFB time.
- Gather's 60 dynamic manifest-type json resources fetched from DB can't be fetched in parallel, they're manifests-of-manifests that need to be recursively fetched; there is interdependency and gather can't fetch all of them in parallel (it has to fetch and parse first, in order to discover what else to fetch).

## Notes

- I chose edutest.temp.build over eskola.tl so that I'd be testing against the same datacenter (both SGP1).
- Load time is wall time spent between clicking login and finishing the loading of resources.
- counting requests: in the asteroid preloading, each request appears twice (page to SW, then SW to web), so the count of 306 needs to be halved (arriving at 153 requests as in the table).
- I've made edutest.temp.build so that it doesn't boot the app after the precaching, if I'd let it boot then it would pollute the network log with stuff that is not the precaching I'm trying to measure. I've left it in this state so it can be used to rerun comparisons if desired

## What to look for in the screenshots
- In the screenshot below it's not well visible because of image scale, but if you look closely you'll see that Canoe is spending time doing... nothing. Obviously using just a single queue is not great, but then leaving it empty while there's stuff to fetch is not helping to further the cause at all
- contrasting Gather and Eskola timelines, it's readily apparent that Eskola does things in parallel, especially in the first half you see that it stuffs the six pipelines, later parallelism is less dense because of resource discovery mentioned earlier (it doesn't know all resources ahead of time).

#### Asteroid devtools network screenshot
![toad](https://user-images.githubusercontent.com/15187983/81938854-09a3ca80-95e5-11ea-8ab6-5485ab0f71eb.png)

#### Gather devtools network screenshot
![hare](https://user-images.githubusercontent.com/15187983/81938870-0e687e80-95e5-11ea-80ee-12c54803a21f.png)

**Comment by PeteCoward** [comment]
So the way I read my latest HAR files ( I'll send you it, format not supported here ) is that 

1. the first four requests are the index,  google analytics, google gtag, and our pre app js.
    - We can reduce the size of our pre-app canoe.js ( and improve naming )

2. on load the service worker then is requested.
   - workbox bloats the size a bit, that might be able to be reduced

3. after the sw load it pre-caches files in parallel, but the big files ( our js and css ) often end up as the only thing being downloaded
   - webpack can chunk them which would help

4. once the sw precaching succeeds, we dynamically load the main app.js, and this should come from the sw cache. I have seen it do so, but have also seen evidence that sometimes it doesn;t and I'm not sure why.
   - needs more investigation

5. After login we pre-load the runtime cache ( the pages come through Django ), and that is sequential because of awaits in our code, I didn't want to swamp the thing as we've seen before and never got around to copying the parallel limiting you did in eskola. It's sequential but should not be holding up the user at that point.
   - could easily add/share your libraries to improve this

**Comment by boerwastaken** [comment]
Yeah it's point 5 that I'm referring to with "preloading resources". What you call the "runtime cache". It takes a minute now for some 30-odd slides. With unreliable networks it's advisable to exploit available connectivity while it lasts, especially if we want to make a good chunk of curriculum available (such as e-learning for science in the Pacific).

Another thing that would ameliorate this (apart from paralellisation) is to bundle up pages for preload, instead of doing a full request cycle roundtrip for every single one. That's also faster on the backend side of things, plus it makes transport compression work better.

**Comment by PeteCoward** [comment]
We have not implemented any transport compression on the JSON page endpoints yet. I figured it would be small gains, compared to the image format work.

**Comment by boerwastaken** [comment]
Yes, responses under a certain size are not worth compressing - nginx even skips small responses.
That's another motivation for instantiation of a bulk endpoint.
